File: src/main/java/io/github/shaksternano/borgar/media/template/ResourceTemplate.java
package io.github.shaksternano.borgar.media.template;

import io.github.shaksternano.borgar.Main;
import io.github.shaksternano.borgar.io.FileUtil;
import io.github.shaksternano.borgar.media.AudioFrame;
import io.github.shaksternano.borgar.media.ImageFrame;
import io.github.shaksternano.borgar.media.ImageUtil;
import io.github.shaksternano.borgar.media.graphics.GraphicsUtil;
import io.github.shaksternano.borgar.media.graphics.Position;
import io.github.shaksternano.borgar.media.graphics.TextAlignment;
import io.github.shaksternano.borgar.media.graphics.drawable.Drawable;
import io.github.shaksternano.borgar.media.io.MediaReaders;
import io.github.shaksternano.borgar.media.io.reader.MediaReader;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Optional;
import java.util.function.Function;

public enum ResourceTemplate implements Template {

    SONIC_SAYS(
        "image/containerimage/sonic_says.png",
        "sonic_says",
        210,
        15,
        620,
        330,
        30,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Bitstream Vera Sans",
        Color.WHITE,
        200,
        null,
        0,
        null,
        true,
        null
    ),

    SOYJAK_POINTING(
        "image/containerimage/soyjak_pointing.png",
        "soyjak_pointing",
        0,
        100,
        1024,
        450,
        0,
        Position.CENTRE,
        250,
        100,
        750,
        450,
        0,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.BLACK,
        200,
        null,
        0,
        null,
        false,
        null
    ),

    THINKING_BUBBLE(
        "image/containerimage/thinking_bubble.png",
        "thinking_bubble",
        12,
        0,
        128,
        81,
        10,
        Position.CENTRE,
        12,
        0,
        128,
        81,
        20,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.BLACK,
        50,
        null,
        0,
        "shape/thinking_bubble_edge_trimmed.javaobject",
        false,
        Color.WHITE
    ),

    MUTA_SOY(
        "image/containerimage/muta_soy.png",
        "muta_soy",
        400,
        256,
        800,
        768,
        0,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.WHITE,
        200,
        null,
        0,
        null,
        false,
        null
    ),

    WALMART_WANTED(
        "image/containerimage/walmart_wanted.png",
        "walmart_wanted",
        428,
        94,
        618,
        318,
        0,
        Position.CENTRE,
        428,
        94,
        618,
        318,
        10,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.BLACK,
        200,
        null,
        0,
        null,
        true,
        null
    ),

    OH_MY_GOODNESS_GRACIOUS(
        "image/containerimage/oh_my_goodness_gracious.gif",
        "oh_my_goodness_gracious",
        250,
        350,
        536,
        640,
        0,
        Position.CENTRE,
        250,
        350,
        536,
        640,
        10,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.BLACK,
        200,
        null,
        0,
        null,
        true,
        null
    ),

    LIVING_IN_1984(
        "image/containerimage/living_in_1984.png",
        "living_in_1984",
        40,
        6,
        350,
        120,
        10,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.BLACK,
        100,
        null,
        0,
        null,
        true,
        null
    ),

    WHO_DID_THIS(
        "image/containerimage/who_did_this.png",
        "wdt",
        0,
        104,
        512,
        403,
        10,
        Position.CENTRE,
        TextAlignment.CENTRE,
        "Futura-CondensedExtraBold",
        Color.BLACK,
        100,
        null,
        0,
        null,
        true,
        Color.WHITE
    ),
    ;

    private final String imagePath;
    private final String format;
    private final String resultName;
    private final int imageContentX;
    private final int imageContentY;
    private final int imageContentWidth;
    private final int imageContentHeight;
    private final Position imageContentPosition;
    private final int textContentX;
    private final int textContentY;
    private final int textContentWidth;
    private final int textContentHeight;
    private final Position textContentPosition;
    private final TextAlignment textContentAlignment;
    private final String fontName;
    private final Font font;
    private final Color textColor;
    @Nullable
    private final Function<String, Drawable> customTextDrawableFactory;
    private final double contentRotationRadians;
    @Nullable
    private final String contentClipShapeFilePath;
    private final boolean isBackground;
    @Nullable
    private final Color fill;

    ResourceTemplate(
        String imagePath,
        String resultName,
        int imageContainerStartX,
        int imageContainerStartY,
        int imageContainerEndX,
        int imageContainerEndY,
        int imageContainerPadding,
        Position imageContentPosition,
        int textContainerStartX,
        int textContainerStartY,
        int textContainerEndX,
        int textContainerEndY,
        int textContainerPadding,
        Position textContentPosition,
        TextAlignment textContentAlignment,
        String fontName,
        Color textColor,
        int maxFontSize,
        @Nullable Function<String, Drawable> customTextDrawableFactory,
        double contentRotationRadians,
        @Nullable String contentClipShapeFilePath,
        boolean isBackground,
        @Nullable Color fill
    ) {
        this.imagePath = imagePath;
        format = determineFormat();
        this.resultName = resultName;
        imageContentX = imageContainerStartX + imageContainerPadding;
        imageContentY = imageContainerStartY + imageContainerPadding;
        textContentX = textContainerStartX + textContainerPadding;
        textContentY = textContainerStartY + textContainerPadding;
        var doubleImagePadding = imageContainerPadding * 2;
        var doubleTextPadding = textContainerPadding * 2;
        imageContentWidth = imageContainerEndX - imageContainerStartX - doubleImagePadding;
        imageContentHeight = imageContainerEndY - imageContainerStartY - doubleImagePadding;
        this.imageContentPosition = imageContentPosition;
        textContentWidth = textContainerEndX - textContainerStartX - doubleTextPadding;
        textContentHeight = textContainerEndY - imageContainerStartY - doubleTextPadding;
        this.textContentPosition = textContentPosition;
        this.textContentAlignment = textContentAlignment;
        this.fontName = fontName;
        font = new Font(this.fontName, Font.PLAIN, maxFontSize);
        this.textColor = textColor;
        this.customTextDrawableFactory = customTextDrawableFactory;
        this.contentRotationRadians = contentRotationRadians;
        this.contentClipShapeFilePath = contentClipShapeFilePath;
        this.isBackground = isBackground;
        this.fill = fill;
    }

    @SuppressWarnings("SameParameterValue")
    ResourceTemplate(
        String imagePath,
        String resultName,
        int contentContainerStartX,
        int contentContainerStartY,
        int contentContainerEndX,
        int contentContainerEndY,
        int contentContainerPadding,
        Position contentPosition,
        TextAlignment textContentAlignment,
        String fontName,
        Color textColor,
        int maxFontSize,
        @Nullable Function<String, Drawable> customTextDrawableFactory,
        double contentRotationRadians,
        @Nullable String contentClipShapeFilePath,
        boolean isBackground,
        @Nullable Color fill
    ) {
        this(
            imagePath,
            resultName,
            contentContainerStartX,
            contentContainerStartY,
            contentContainerEndX,
            contentContainerEndY,
            contentContainerPadding,
            contentPosition,
            contentContainerStartX,
            contentContainerStartY,
            contentContainerEndX,
            contentContainerEndY,
            contentContainerPadding,
            contentPosition,
            textContentAlignment, fontName,
            textColor,
            maxFontSize,
            customTextDrawableFactory,
            contentRotationRadians,
            contentClipShapeFilePath,
            isBackground,
            fill
        );
    }

    private String determineFormat() {
        try (var inputStream = FileUtil.getResourceInRootPackage(imagePath)) {
            return ImageUtil.getImageFormat(inputStream);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public MediaReader<ImageFrame> getImageReader() throws IOException {
        var inputStream = FileUtil.getResourceInRootPackage(imagePath);
        return MediaReaders.createImageReader(inputStream, format);
    }

    @Override
    public MediaReader<AudioFrame> getAudioReader() throws IOException {
        var inputStream = FileUtil.getResourceInRootPackage(imagePath);
        return MediaReaders.createAudioReader(inputStream, format);
    }

    @Override
    public String getFormat() {
        return format;
    }

    @Override
    public String getResultName() {
        return resultName;
    }

    @Override
    public int getImageContentX() {
        return imageContentX;
    }

    @Override
    public int getImageContentY() {
        return imageContentY;
    }

    @Override
    public int getImageContentWidth() {
        return imageContentWidth;
    }

    @Override
    public int getImageContentHeight() {
        return imageContentHeight;
    }

    @Override
    public Position getImageContentPosition() {
        return imageContentPosition;
    }

    @Override
    public int getTextContentX() {
        return textContentX;
    }

    @Override
    public int getTextContentY() {
        return textContentY;
    }

    @Override
    public int getTextContentWidth() {
        return textContentWidth;
    }

    @Override
    public int getTextContentHeight() {
        return textContentHeight;
    }

    @Override
    public Position getTextContentPosition() {
        return textContentPosition;
    }

    @Override
    public TextAlignment getTextContentAlignment() {
        return textContentAlignment;
    }

    @Override
    public Font getFont() {
        return font;
    }

    @Override
    public Color getTextColor() {
        return textColor;
    }

    @Override
    public Optional<Function<String, Drawable>> getCustomTextDrawableFactory() {
        return Optional.ofNullable(customTextDrawableFactory);
    }

    @Override
    public double getContentRotation() {
        return contentRotationRadians;
    }

    @Override
    public Optional<Shape> getContentClip() throws IOException {
        if (contentClipShapeFilePath == null) {
            return Optional.empty();
        } else {
            return Optional.of(GraphicsUtil.loadShape(contentClipShapeFilePath));
        }
    }

    @Override
    public boolean isBackground() {
        return isBackground;
    }

    @Override
    public Optional<Color> getFill() {
        return Optional.ofNullable(fill);
    }

    public static void validate() {
        for (var containerImageInfo : ResourceTemplate.values()) {
            validateImage(containerImageInfo);
            validateFont(containerImageInfo);
            validateContentClip(containerImageInfo);
        }
    }

    private static void validateImage(ResourceTemplate containerImageInfo) {
        if (containerImageInfo.imagePath == null) {
            Main.getLogger().error("Image resource path in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\" is null!");
        } else {
            try {
                FileUtil.validateResourcePathInRootPackage(containerImageInfo.imagePath);

                try (
                    @SuppressWarnings("unused")
                    var imageReader = containerImageInfo.getImageReader();
                    @SuppressWarnings("unused")
                    var audioReader = containerImageInfo.getAudioReader()
                ) {
                    return;
                } catch (Throwable t) {
                    Main.getLogger().error("Error loading media with path \"" + containerImageInfo.imagePath + "\" in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\"!", t);
                }
            } catch (Throwable t) {
                Main.getLogger().error("Media resource path \"" + containerImageInfo.imagePath + "\" in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\" is invalid!", t);
            }
        }

        Main.shutdown(1);
    }

    private static void validateFont(ResourceTemplate containerImageInfo) {
        if (containerImageInfo.fontName == null) {
            Main.getLogger().error("Font name in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\" is null!");
        } else {
            var font = new Font(containerImageInfo.fontName, Font.PLAIN, 1);
            if (font.getFontName().equals(containerImageInfo.fontName)) {
                return;
            } else {
                Main.getLogger().error("Font \"" + containerImageInfo.fontName + "\" in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\" is not a registered font!");
            }
        }

        Main.shutdown(1);
    }

    private static void validateContentClip(ResourceTemplate containerImageInfo) {
        if (containerImageInfo.contentClipShapeFilePath != null) {
            try {
                FileUtil.validateResourcePathInRootPackage(containerImageInfo.contentClipShapeFilePath);

                try {
                    GraphicsUtil.loadShape(containerImageInfo.contentClipShapeFilePath);
                    return;
                } catch (Throwable t) {
                    Main.getLogger().error("Error loading shape with path \"" + containerImageInfo.contentClipShapeFilePath + "\" in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\"!", t);
                }
            } catch (Throwable t) {
                Main.getLogger().error("Shape resource path \"" + containerImageInfo.contentClipShapeFilePath + "\" in " + containerImageInfo.getClass().getSimpleName() + " \"" + containerImageInfo + "\" is invalid!", t);
            }

            Main.shutdown(1);
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/template/ResourceTemplate.kt
package io.github.shaksternano.borgar.core.media.template

import io.github.shaksternano.borgar.core.graphics.ContentPosition
import io.github.shaksternano.borgar.core.graphics.TextAlignment
import io.github.shaksternano.borgar.core.graphics.drawable.Drawable
import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.io.fileExtension
import io.github.shaksternano.borgar.core.media.createAudioReader
import io.github.shaksternano.borgar.core.media.createImageReader
import io.github.shaksternano.borgar.core.media.reader.AudioReader
import io.github.shaksternano.borgar.core.media.reader.ImageReader
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.awt.Color
import java.awt.Font
import java.awt.Shape
import java.io.IOException
import java.io.ObjectInputStream

enum class ResourceTemplate(
    private val mediaPath: String,
    override val resultName: String,

    imageContainerStartX: Int,
    imageContainerStartY: Int,
    imageContainerEndX: Int,
    imageContainerEndY: Int,
    imageContainerPadding: Int,
    override val imageContentPosition: ContentPosition,

    textContainerStartX: Int,
    textContainerStartY: Int,
    textContainerEndX: Int,
    textContainerEndY: Int,
    textContainerPadding: Int,
    override val textContentPosition: ContentPosition,
    override val textContentAlignment: TextAlignment,

    fontName: String,
    maxFontSize: Int,
    override val textColor: Color,
    override val isBackground: Boolean,
    override val contentRotationRadians: Double = 0.0,
    override val fill: Color? = null,
    private val contentClipShapePath: String? = null,
    override val customTextDrawableSupplier: ((String) -> Drawable)? = null,
    override val forceTransparency: Boolean = false,
) : Template {

    SONIC_SAYS(
        mediaPath = "media/containerimage/sonic_says.png",
        resultName = "sonic_says",

        contentContainerStartX = 210,
        contentContainerStartY = 15,
        contentContainerEndX = 620,
        contentContainerEndY = 330,
        contentContainerPadding = 30,
        contentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Bitstream Vera Sans",
        maxFontSize = 200,
        textColor = Color.WHITE,
        isBackground = true,
        forceTransparency = true,
    ),

    SOYJAK_POINTING(
        mediaPath = "media/containerimage/soyjak_pointing.png",
        resultName = "soyjak_pointing",

        imageContainerStartX = 0,
        imageContainerStartY = 100,
        imageContainerEndX = 1024,
        imageContainerEndY = 450,
        imageContainerPadding = 0,
        imageContentPosition = ContentPosition.CENTRE,

        textContainerStartX = 250,
        textContainerStartY = 100,
        textContainerEndX = 750,
        textContainerEndY = 450,
        textContainerPadding = 0,
        textContentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 200,
        textColor = Color.BLACK,
        isBackground = false,
        forceTransparency = true,
    ),

    THINKING_BUBBLE(
        mediaPath = "media/containerimage/thinking_bubble.png",
        resultName = "thinking_bubble",

        imageContainerStartX = 12,
        imageContainerStartY = 0,
        imageContainerEndX = 128,
        imageContainerEndY = 81,
        imageContainerPadding = 10,
        imageContentPosition = ContentPosition.CENTRE,

        textContainerStartX = 12,
        textContainerStartY = 0,
        textContainerEndX = 128,
        textContainerEndY = 81,
        textContainerPadding = 20,
        textContentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 50,
        textColor = Color.BLACK,
        isBackground = false,
        fill = Color.WHITE,
        contentClipShapePath = "shape/thinking_bubble_edge_trimmed.javaobject",
        forceTransparency = true,
    ),

    MUTA_SOY(
        mediaPath = "media/containerimage/muta_soy.png",
        resultName = "muta_soy",

        contentContainerStartX = 400,
        contentContainerStartY = 256,
        contentContainerEndX = 800,
        contentContainerEndY = 768,
        contentContainerPadding = 0,
        contentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 100,
        textColor = Color.WHITE,
        isBackground = false,
        forceTransparency = true,
    ),

    WALMART_WANTED(
        mediaPath = "media/containerimage/walmart_wanted.png",
        resultName = "walmart_wanted",

        imageContainerStartX = 428,
        imageContainerStartY = 94,
        imageContainerEndX = 618,
        imageContainerEndY = 318,
        imageContainerPadding = 0,
        imageContentPosition = ContentPosition.CENTRE,

        textContainerStartX = 428,
        textContainerStartY = 94,
        textContainerEndX = 618,
        textContainerEndY = 318,
        textContainerPadding = 10,
        textContentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 200,
        textColor = Color.BLACK,
        isBackground = true,
    ),

    OH_MY_GOODNESS_GRACIOUS(
        mediaPath = "media/containerimage/oh_my_goodness_gracious.gif",
        resultName = "oh_my_goodness_gracious",

        imageContainerStartX = 250,
        imageContainerStartY = 350,
        imageContainerEndX = 536,
        imageContainerEndY = 640,
        imageContainerPadding = 0,
        imageContentPosition = ContentPosition.CENTRE,

        textContainerStartX = 250,
        textContainerStartY = 350,
        textContainerEndX = 536,
        textContainerEndY = 640,
        textContainerPadding = 10,
        textContentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 200,
        textColor = Color.BLACK,
        isBackground = true,
    ),

    LIVING_IN_1984(
        mediaPath = "media/containerimage/living_in_1984.png",
        resultName = "living_in_1984",

        contentContainerStartX = 40,
        contentContainerStartY = 6,
        contentContainerEndX = 350,
        contentContainerEndY = 120,
        contentContainerPadding = 10,
        contentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 100,
        textColor = Color.BLACK,
        isBackground = true,
    ),

    WHO_DID_THIS(
        mediaPath = "media/containerimage/who_did_this.png",
        resultName = "who_did_this",

        contentContainerStartX = 0,
        contentContainerStartY = 104,
        contentContainerEndX = 512,
        contentContainerEndY = 403,
        contentContainerPadding = 10,
        contentPosition = ContentPosition.CENTRE,
        textContentAlignment = TextAlignment.CENTRE,

        fontName = "Futura-CondensedExtraBold",
        maxFontSize = 100,
        textColor = Color.BLACK,
        isBackground = true,
        fill = Color.WHITE,
    ),
    ;

    constructor(
        mediaPath: String,
        resultName: String,
        contentContainerStartX: Int,
        contentContainerStartY: Int,
        contentContainerEndX: Int,
        contentContainerEndY: Int,
        contentContainerPadding: Int,
        contentPosition: ContentPosition,
        textContentAlignment: TextAlignment,
        fontName: String,
        maxFontSize: Int,
        textColor: Color,
        isBackground: Boolean,
        contentRotationRadians: Double = 0.0,
        fill: Color? = null,
        contentClipShapePath: String? = null,
        customTextDrawableSupplier: ((String) -> Drawable)? = null,
        forceTransparency: Boolean = false,
    ) : this(
        mediaPath,
        resultName,

        contentContainerStartX,
        contentContainerStartY,
        contentContainerEndX,
        contentContainerEndY,
        contentContainerPadding,
        contentPosition,

        contentContainerStartX,
        contentContainerStartY,
        contentContainerEndX,
        contentContainerEndY,
        contentContainerPadding,
        contentPosition,
        textContentAlignment,

        fontName,
        maxFontSize,
        textColor,
        isBackground,
        contentRotationRadians,
        fill,
        contentClipShapePath,
        customTextDrawableSupplier,
        forceTransparency,
    )

    override val media: DataSource = DataSource.fromResource(mediaPath)
    override val format: String = fileExtension(mediaPath)

    override val imageContentX: Int = imageContainerStartX + imageContainerPadding
    override val imageContentY: Int = imageContainerStartY + imageContainerPadding
    override val imageContentWidth: Int = imageContainerEndX - imageContainerStartX - 2 * imageContainerPadding
    override val imageContentHeight: Int = imageContainerEndY - imageContainerStartY - 2 * imageContainerPadding

    override val textContentX: Int = textContainerStartX + textContainerPadding
    override val textContentY: Int = textContainerStartY + textContainerPadding
    override val textContentWidth: Int = textContainerEndX - textContainerStartX - 2 * textContainerPadding
    override val textContentHeight: Int = textContainerEndY - textContainerStartY - 2 * textContainerPadding

    override val font: Font = Font(fontName, Font.PLAIN, maxFontSize)

    override suspend fun getImageReader(): ImageReader {
        val dataSource = DataSource.fromResource(mediaPath)
        return createImageReader(dataSource)
    }

    override suspend fun getAudioReader(): AudioReader {
        val dataSource = DataSource.fromResource(mediaPath)
        return createAudioReader(dataSource)
    }

    override suspend fun getContentClip(): Shape? {
        if (contentClipShapePath == null) return null
        val dataSource = DataSource.fromResource(contentClipShapePath)
        return runCatching {
            val inputStream = dataSource.newStream()
            val parsedObject = withContext(Dispatchers.IO) {
                ObjectInputStream(inputStream).use {
                    it.readObject()
                }
            }
            if (parsedObject is Shape) parsedObject
            else {
                var errorMessage = "The parsed object under $contentClipShapePath is not a shape"
                val parsedObjectClassName = parsedObject::class.qualifiedName
                if (parsedObjectClassName != null)
                    errorMessage += ", but a $parsedObjectClassName"
                throw IOException(errorMessage)
            }
        }.getOrElse {
            throw IOException("Failed to load shape file under $contentClipShapePath", it)
        }
    }
}
